### PR TITLE
Fix(49472): Added docs for Set and Map types

### LINKED ISSUE #49472: No documentation for Map and Set types
# lib Update Request

## Configuration Check

My compilation *target* is `ES2020` and my *lib* is `the default`.

## Missing / Incorrect Definition
`Set` and `Map`

## Sample Code

The `Map` and `Set` types don't appear to have any documentation. This makes it difficult for a beginner to work with them

```ts
const x = new Set();
x.clear() // no docs here
```

## Documentation Link

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

### PR DESCRIPTION
**Pull request by Grubba27**
Fixes #49472

### CODE REVIEW

**Review by sandersn — COMMENTED** [review]
Reading the rest of the comments, I feel the same way about nearly all of them: Some parts are valuable, the rest is redundant. @DanielRosenwasser or @andrewbranch what do you think about incomplete-but-interesting JSDoc vs complete-but-redundant?

File: src/lib/es2015.collection.d.ts [hunk 1]
```diff
@@ -1,16 +1,47 @@
...
+    /**
+     * Returns a boolean indicating whether an element with the specified key exists or not.
+     * @param key The key of the element to test for presence in the Map object.
+     * @returns Returns true if an element in the Map object exists with the specified key, or false otherwise.
+     */
     has(key: K): boolean;
+    /**
+     * Adds a new element with a specified key and value to the Map object. If an element with the same key already exists, the element will be updated.
+     * @param key The key of the element to add to the Map object. The key may be any JavaScript type (any primitive value or any type of JavaScript object).
+     * @param value The value of the element to add to the Map object. The value may be any JavaScript type (any primitive value or any type of JavaScript object).
```
[the remark this answers is not in the document]
Comment (Grubba27) [reply src/lib/es2015.collection.d.ts:33]: I think almost the same. Most of these are MDN docs with some adjustments to make them easier to read while you are coding. Without the need for going to the docs. But if it makes more sense I would be delighted to adjust

[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/lib/es2015.collection.d.ts]: Personally I would delete any JSDoc that essentially just rewrites the code (or other in-scope JSDoc comments) as an English sentence. E.g.

> `@param key The key of the element to remove from the Map object.`

- `The key` literally just the name of the parameter
- `of the element` I think that’s implied
- `to remove` just a synonym for `delete` which is the name of the method
- `from the Map` just the name of the type
- `object` c'mon 😄

**Review by sandersn — CHANGES_REQUESTED** [review]
Please remove all the redundant comments based on @andrewbranch 's criterion.

**Comment by Grubba27** [comment]
Okay! I've done some adjusts with the jsdocs but I'm still not sure if they are all right especially with these two parts:
```ts
    /**
     * Returns a specified element from the Map object. If the value that is associated with the provided key is an object, then you will get a reference to that object and any change made to that object will effectively modify it inside the Map.
     * @param key
     * @returns Returns the element associated with the specified key. If no element is associated with the specified key, undefined is returned.
     */
    get(key: K): V | undefined;
```

makes sense to have the description and the return part?

also, my other question would be if makes sense having jsdocs like this one:
```ts
    /**
     * Removes the specified element.
     * @param key
     */
    delete(key: K): boolean;
```
because the method is already described itself. I'm a little bit confused

**Comment by andrewbranch** [comment]
1. I think you can just delete `@param key` if it doesn’t have a description.
2. For self-explanatory methods like `delete`, I don’t think it really needs the description, but I don’t have a strong opinion. However, `delete` should definitely have a `@returns` since it’s not at all obvious what that boolean means. I think it might read a little weird to have `@param` or `@returns` tags with no top-level description, so I’d be ok with leaving `Removes the specified element` in. @sandersn what do you think?

File: src/lib/es2015.collection.d.ts [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
+
     clear(): void;
+    /**
+     * @returns true if an element in the Map existed and has been removed, or false if the element does not exist.
+     */
     delete(key: K): boolean;
+    /**
+     * Executes a provided function once per each key/value pair in the Map, in insertion order.
+     * @param callbackfn
+     * @param thisArg
```
Comment (andrewbranch) [inline src/lib/es2015.collection.d.ts:11]: `@param`s without a description don’t do anything, I’m pretty sure (same thing on `Set`)

**Review by andrewbranch — COMMENTED** [review]
A couple straggling `@params` but otherwise I think this looks good

**Review by andrewbranch — APPROVED** [review]
Thanks @Grubba27!

**Review by sandersn — APPROVED** [review]

**Pull request merged**